// artemisa/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _component = require('./lib/artemisa/component');

Object.defineProperty(exports, 'fetchingData', {
  enumerable: true,
  get: function get() {
    return _component.fetchingData;
  }
});

var _reducer = require('./lib/artemisa/reducer');

Object.defineProperty(exports, 'reducer', {
  enumerable: true,
  get: function get() {
    return _reducer.artemisa;
  }
});
Object.defineProperty(exports, 'isFetching', {
  enumerable: true,
  get: function get() {
    return _reducer.isFetching;
  }
});
Object.defineProperty(exports, 'isFetched', {
  enumerable: true,
  get: function get() {
    return _reducer.isFetched;
  }
});
Object.defineProperty(exports, 'processApiAction', {
  enumerable: true,
  get: function get() {
    return _reducer.processApiAction;
  }
});

var _service = require('./lib/core/service');

Object.defineProperty(exports, 'service', {
  enumerable: true,
  get: function get() {
    return _service.dataService;
  }
});

var _model = require('./lib/core/model');

Object.defineProperty(exports, 'State', {
  enumerable: true,
  get: function get() {
    return _model.State;
  }
});
Object.defineProperty(exports, 'isErrorSlot', {
  enumerable: true,
  get: function get() {
    return _model.isErrorSlot;
  }
});
Object.defineProperty(exports, 'isFetchedSlot', {
  enumerable: true,
  get: function get() {
    return _model.isFetchedSlot;
  }
});
Object.defineProperty(exports, 'isFetchingSlot', {
  enumerable: true,
  get: function get() {
    return _model.isFetchingSlot;
  }
});

var _dispatch = require('./lib/artemisa/dispatch');

Object.defineProperty(exports, 'shouldFetch', {
  enumerable: true,
  get: function get() {
    return _dispatch.shouldFetch;
  }
});
Object.defineProperty(exports, 'actionName', {
  enumerable: true,
  get: function get() {
    return _dispatch.actionName;
  }
});

var _actions = require('./lib/core/actions');

Object.defineProperty(exports, 'simulateDataReceive', {
  enumerable: true,
  get: function get() {
    return _actions.simulateDataReceive;
  }
});
Object.defineProperty(exports, 'isDerivedActionFor', {
  enumerable: true,
  get: function get() {
    return _actions.isDerivedActionFor;
  }
});
Object.defineProperty(exports, 'receiveTypeFor', {
  enumerable: true,
  get: function get() {
    return _actions.receiveTypeFor;
  }
});
